### ISSUE TITLE
Missing SiteLogoUrl and other props under O365 16.1 runtime

### ISSUE THREAD

**Issue by adaoust** [SubPointSolutions/spmeta2#939]
In the RootWebDefinition, there is not support for changing or setting the SiteLogoUrl, which means that from a SiteModel it is impossible to set. Some other properties that are in the WebDefinition are also missing from the RootWebDefinition.

**Comment by SubPointSupport** [comment]
Hey @adaoust, thanks for reporting this,

RootWebDefinition was introduced to "lookup" the root web from the "web models" level. It was designed to deploy some artifacts to the root web (such as style library) while "being within web model - subweb model".

For your case, you may consider using WebDefinition with "Url" set to "/". Add such definition under SiteModel so that you'd have web with "/" Url. If no Title/Description set - that's fine, no changes will be done but the rest of the pros will still be applied - such as SiteLogoUrl and the rest.

Let us know what you think, and if there is anything else we can help you with.
Also, don't forget to check out Yammer Network as well as docs site:
* [SPMeta2 Yammer Network](https://www.yammer.com/spmeta2feedback/)
* [SPMeta2 docs](http://docs.subpointsolutions.com/spmeta2)
* [SPMeta2 scenarios](http://docs.subpointsolutions.com/spmeta2/scenarios)

**Comment by adaoust** [comment]
Great I didn't know I could use "/" for the RootWeb!

Thanks again for the great support!

**Comment by SubPointSupport** [comment]
Sure, no worries. Give it a go, let us know how it works for you.

**Comment by adaoust** [comment]
Hey guys,

I could not get it to work (in latest release or beta4).

Here are a few things I tried.

**Attempt 1.**
```
site.AddHostWeb(new WebDefinition()
{
    Url = "/",
    SiteLogoUrl = "/SiteAssets/LogoTest.png"
});
```

 => Nothing happens because the node is not processed.

**Attempt 2.**
```
site.AddWeb(new WebDefinition()
{
    Url = "/",
    SiteLogoUrl = "/SiteAssets/LogoTest.png"
});
```

 => Errors while validating the model
 => Inner Exception 1: Property [Title] is not valid.
 => Inner Exception 2 One of the properties with [Web Template] attribute should be set. Definition:[Title:[] Description:[] Convert(p.LCID):[1033] Convert(p.UseUniquePermission):[False] Url:[/] WebTemplate:[] CustomWebTemplate:[]]

 FYI, I also noticed the SiteLogoUrl is not in the ToString() of the WebDefinition. However, it is correctly set in memory.

**Attempt 3.**
```
site.AddWeb(new WebDefinition()
{
    Title = "Accessibility", // Same as site
    WebTemplate = BuiltInWebTemplates.Collaboration.TeamSite, // Same as site
    Url = "/",
    SiteLogoUrl = "/SiteAssets/LogoTest.png"
});
```

 => There was an error while provisioning definition. Check ModelNode prop.
 => Inner Exception: Method "SiteLogoUrl" does not exist.

Let me know if you can help!

Thanks,

Alex

**Comment by SubPointSupport** [comment]
Okay, we'll have a closer look. Just to confirm the desired case - update the SiteLogoUrl property of the existing root web, righ?

**Comment by adaoust** [comment]
Yes exactly, thanks!

**Comment by SubPointSupport** [comment]
Hello @adaoust, 

We had a closer look, here how it all works: 

Attempt 3 is a correct one. SPMeta2 creates or updates SharePoint artifacts as per well-formed definitions. Unfortionarly, we don't have per-property update, so that it means that we need to for a correct WebDefinition with all "minimal" properties such as Url, Title, Description and so on. SPMeta2 uses .Url to understand if web already exists, then creates a new web (if it didn't exist), and then updates properties as they are set in the definition. In turn, all "writable" properties will be overwritten.

```cs
site.AddWeb(new WebDefinition()
{
    Title = "Accessibility", // Same as site
    WebTemplate = BuiltInWebTemplates.Collaboration.TeamSite, // Same as site
    Url = "/",
    SiteLogoUrl = "/SiteAssets/LogoTest.png"
});
```

At this point, the model looks correct and that's exactly the right way to go.

Next, the exception:
```
=> Inner Exception: Method "SiteLogoUrl" does not exist.
```
There might be a few reasons to have such exception:
* With SSOM - old SPMeta2 assemblies in the GAC (so that definition actually doesn't have that prop)
* With CSOM + onprem - SharePoint 2013 CSOM is old (farm is below the required CSOM runtome) and the method does not exist on SharePoint farm at all, or the CSOM assemblies are old (solution has to refernece new CSOM assemblies)
* With CSOM + O365 - either O365 tenant is old, or the CSOM assemblies are old (solution has to refernece new CSOM assemblies)

Let us know what SharePoint version and API is used, and then what is the outcome of the following SPMeta2 call. Will take it from there:
```cs
var info = SPMeta2Diagnostic.GetDiagnosticInfo();
```

**Comment by adaoust** [comment]
Hi @SubPointSupport, 

I am running my code against an O365 Tenant (actually tried it on a few different ones). I stripped the code down to the strict minimum and I still get the same issue.

Here's the code I am running:

And here's the Output:

**Comment by SubPointSupport** [comment]
Alright, CSOM assembly seems to be the latest (16.1.6112.1200), so in that case we would think that the issue is an "old" tenant.

ClientContext has the server version (the server shema version or something) - that's the first point of interest. Next, try to update the web with pure C# code (update SiteLogoUrl with the pure C# code) - you won't be able to, having the same issue on missed property.

Most likely, the O365 tenant is older than the actual CSOM assemblies and really does not have that prop. You'd need to ask MS to update the tenant to the latest version.

**Comment by adaoust** [comment]
Hi @SubPointSupport,

Setting the Site Logo works perfectly fine with pure C#/CSOM code.

Here's some code that works: 

 ```
static void Main(string[] args)
        {
            Console.Write("Enter your site url: ");
            string url = Console.ReadLine();

            Console.Write("Enter your username: ");
            string username = Console.ReadLine();

            SecureString password = ReadPassword();

            try
            {
                using (ClientContext clientContext = new ClientContext(url))
                {
                    clientContext.Credentials = new SharePointOnlineCredentials(username, password);
                    clientContext.Load(clientContext.Web);
                    clientContext.ExecuteQuery();

                    clientContext.Site.RootWeb.SiteLogoUrl = "/SiteAssets/LogoTest.png";
                    clientContext.Site.RootWeb.Update();
                    clientContext.ExecuteQuery();
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: " + ex.Message);
            }

            // Keep Console Open.
            if (Debugger.IsAttached) Console.ReadKey();
        }

        private static SecureString ReadPassword()
        {
            SecureString securePassword = new SecureString();
            Console.Write("Enter your password: ");
            ConsoleKeyInfo key;
            do
            {
                key = Console.ReadKey(true);
                // Backspace Should Not Work
                if (key.Key != ConsoleKey.Backspace && key.Key != ConsoleKey.Enter)
                {
                    securePassword.AppendChar(key.KeyChar);
                    Console.Write("*");
                }
                else if (key.Key != ConsoleKey.Enter) Console.Write("\b");
            }
            while (key.Key != ConsoleKey.Enter);
            Console.WriteLine();
            return securePassword;
        }
```

I looked at the CSOM WebModelHandler in the MapProperties method and it seems to be trying to call SiteLogoUrl as a Method instead of just setting the property. I know this is done to avoid problems with older assemblies, but it might be causing the issue. 

```
context.AddQuery(new ClientActionInvokeMethod(web, "SiteLogoUrl", new object[]
{
        webModel.SiteLogoUrl
}));
```

Maybe it should be setting the property directly instead. I find it weird that this would have ever worked.

Thanks again,

Alex

**Comment by SubPointSupport** [comment]
Bug, then. Looking into it this week. Thank you for your time and patience.

**Comment by adaoust** [comment]
Hi @SubPointSupport,

In my last message, I noted that the `WebModelHandler` might be the issue and confirmed this myself by testing it out. 

The following code throws the "Method "SiteLogoUrl" does not exist." exception.

`context.AddQuery(new ClientActionInvokeMethod(web, "SiteLogoUrl", new object[] { webModel.SiteLogoUrl }));`

But if I change it to:

`context.AddQuery(new ClientActionSetProperty(web, "SiteLogoUrl", webModel.SiteLogoUrl);`

It works perfectly.

Not sure if a change was made to `ClientActionInvokeMethod` or if it is used incorrectly, but it for sure doesn't work for setting properties with the current versions of the assemblies. I noticed that `ClientActionInvokeMethod` is used in multiple places, it might be helpful to confirm it works for those.

Hope that helps, I would create a Pull Request, but I couldn't find instructions on how to do so for you guys.

Thanks,

Alex

**Comment by SubPointSupport** [comment]
Hey @adaoust, 

Thanks for getting into details. Confirming that all findings are valid. ClientActionInvokeMethod was nitially used for setting up localization (which is actually a method), and then extrapolated to other properties (which is totally wrong!). As you wrote, ClientActionSetProperty should be used for setting up props.

We are inspecting both M2 source code and client assemblies to get it right. Will test manually for the time being ensuring all works. Full O365 testing under 16.1 asemblies requires some engineering effort, it'll come later.

Again, really appreciate you looked into that. Updates will be avialable in nuget this week.

**Comment by SubPointSupport** [comment]
Fixes for O365 and v16.1 runtime. 

Regression testing was rewritten to identify and load correct O365 runtime as well as all SPMeta2 v16 assemblies - CSOM and Regression.CSOM implementations. 

SPMeta2.CSOM API has a new service to handle CSOM props/methods calls via reflection. All related code was cleanup and refactored so that all reflection based calls come to ClientRuntimeQueryServiceBase instance. Default implementation is DefaultClientRuntimeQueryService, it can be changed with your own implementation via service container registration:
```
public class MyService : ClientRuntimeQueryServiceBase {

}
 ServiceContainer.Instance.ReplaceService(typeof(ClientRuntimeQueryServiceBase ), new MyService ());
```
By default, DefaultClientRuntimeQueryService is used.

While refactoring, several typos were identified (props setup). Fixed, regression testing was done for WebDefinition and some other artifacts that use reflection to set props/methods. Confirming that all works well for WebDefinition. Further SP2013 regression is to be done before releasing the final package to NuGet.

Closing this ticket as the rest is matter of running full regression for SP2013/O365 and ensuring all tests are green.
